File: Java/OOP February - April 2021/05. Polymorphisum/E02VehiclesExtension/Truck.java
package E02VehiclesExtension;

public class Truck extends Vehicle {


    public Truck(double fuelQuantity, double consumptionLitersPerKilometer, double tankCapacity) {
        super(fuelQuantity, consumptionLitersPerKilometer, tankCapacity);
    }

    @Override
    public void setConsumptionLitersPerKilometer(double consumptionLitersPerKilometer) {
        super.setConsumptionLitersPerKilometer(consumptionLitersPerKilometer + 1.6);
    }

    @Override
    public void refuel(double fuel) {
        super.refuel(fuel * 0.95);
    }

    @Override
    public String drive(double distance) {
        return "Truck " + super.drive(distance);
    }
}

File: C#/OOP/04_Polymorphisum_Exercise/E02_VehiclesExtension/Models/Truck.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace VehiclesExtension.Models
{
    public class Truck : Vehicle
    {
        private const double IncreasedConsumption = 1.6;

        public Truck(double fuelQuantity, double fuelConsumption, double tankCapacity)
            : base(fuelQuantity, fuelConsumption, tankCapacity, IncreasedConsumption)
        {
        }

        public override void Refuel(double fuelAmount)
        {
            if (fuelAmount + FuelQuantity > TankCapacity)
            {
                throw new ArgumentException($"Cannot fit {fuelAmount} fuel in the tank");
            }

            base.Refuel(fuelAmount * 0.95);
        }
    }
}
